[BSC] Proposed implementation for issue8 (NetCDF attributes in FieldsIO)

andreapiacentini opened the pull request:
Here comes an AI drafted implementation for dynamically recovering the attributes and writing them out.  
I've added my two cents in making them go through the State and Increment methods.

Fixes #8.

andreapiacentini committed:
[BSC] Proposed implementation for issue8 (NetCDF attributes in FieldsIO)

benjaminmenetrier committed:
Merge branch 'develop' into feature/netcdf-attributes

benjaminmenetrier committed:
Use fields metadata

benjaminmenetrier commented:
Thank you @andreapiacentini. I pushed a commit with a slightly different strategy:
- The attributes are now stored in `Fields` metadata (`fset_[varName].metadata()`), which is an `eckit::LocalConfiguation`. It is more general than the attributes structure you proposed, and is already used for other properties. That way, we can keep the `Fields` class more generic.
- However, I moved your attributes structure into `FieldsIOBSC`, with dedicated private methods.

I've tested the convertstate, and attributes seem to be correctly read and written. Please check other tests and let me know what you think about this implementation.

andreapiacentini commented:
I've lost the attributes in the Dirac output and more in general in all increment outputs.  
I guess the replacement on test on `varAttributes_.empty()` by  `metadata.empty()` is not working.

andreapiacentini commented:
I see that attributes are stored here: 
`metadata.set("attributes", attrsConfig);`  

The test should become something like (I am not able to write it directly in C++, I do somethins pythonlike
```
hasAttr = False
if "attributes" in metadata:
    hasAttr = !metadata["attributes"].empty()
```
Whats the C++ syntax ?

B.t.w. have you got an opinion about why `isState_` is not copied in the `=` operator?

andreapiacentini commented:
Could it be something like that ? (A bit verbose)
I've added my `//AP` comments where I have changed things or where I have questions.
I let it to you, because my afternoon is a bit busy, sorry.
```
    // Copy fieldset
    fsetRhs = util::copyFieldSet(rhsInterp.fset_);
  }
  //AP Added once, ok ??
  eckit::LocalConfiguration attrsConfig;
  for (const auto & var : vars_) {
[...]
        }
        //AP Check if field has attributes and they are not empty
        field_has_attr = field.metadata().get("attributes", attrsConfig);
        if (field_has_attr) {
          field_has_attr = !attrsConfig.empty();
        }
        //AP Check if rhs has attributes and they are not empty
        rhs_has_attr = fieldRhs.metadata().get("attributes", attrsConfig);
        if (rhs_has_attr) {
          rhs_has_attr = !attrsConfig.empty();
        }  
        //AP Copying the whole metadata even if only attributes are missing        
        if (!field_has_attr && rhs_has_attr) {
          field.metadata() = fieldRhs.metadata();
        }
        field.set_dirty(field.dirty() || fieldRhs.dirty());
      }
    }  else {
```

benjaminmenetrier commented:
You are right, I'm pushing a fix in a few minutes.

andreapiacentini commented:
You are doing certainly better. I guess that if my way was the way to go, the whole get + empty test could become a method of Fields

benjaminmenetrier committed:
Uses util::mergeConfigs for metadata

benjaminmenetrier commented:
I want to use the `util::mergeConfigs` tool from OOPS (`oops/src/oops/util/ConfigFunctions.h`), but it needs a minor update to work in our case. I've done this in a new OOPS branch, please use the following line in your bundle `CMakeLists.txt`:
```
ecbuild_bundle( PROJECT oops GIT "https://github.com/benjaminmenetrier/oops.git" BRANCH feature/mergeConfigsBehavior UPDATE )
```

I'm opening a PR now to have it merge as soon as possible in JCSDA `develop`.

andreapiacentini commented:
I fear this branch is private

benjaminmenetrier commented:
You are right. It is also on `jcsda-internal`, do you have access to it?

andreapiacentini commented:
Yes, I have. Yet, if we want it installed at BSC it should either be on official JCSDA branches or on public branches of one of your clones.

benjaminmenetrier commented:
You (and BSC people) can use the following public OOPS repo:
```
ecbuild_bundle( PROJECT oops GIT "https://github.com/benjaminmenetrier/oops_public.git" BRANCH feature/mergeConfigsBehavior UPDATE )
```

andreapiacentini commented:
We are not there yet. I check the files which are ok:  
- [x] monarch_state_out.nc
- [x] monarch_output.nc
- [x] monarch_levsel_output.nc
- [x] monarch_levsel_scaling_output.nc
- [x] monarch_levsel_logtransf_output.nc
- [x] monarch_mean.nc
- [x] monarch_variance.nc
- [x] monarch_sttdev.nc
- [x] monarch_D3reg_no_loc.nc
- [x] monarch_D3reg_no_loc_lt.nc
- [ ] monarch_D4IDreg_2022070104.nc
- [ ] monarch_D4IDreg_2022070103.nc
- [ ] monarch_D4IDreg_2022070101.nc
- [ ] monarch_D4IDreg_2022070100.nc
- [ ] monarch_D4IDreg_2022070105.nc
- [ ] monarch_D4IDreg_2022070102.nc
- [x] monarch_D4reg_no_loc_20220701.nc
- [x] monarch_D3reg_ensemble.nc
- [ ] monarch_D3reg_ensemble_localization.nc
- [x] monarch_D3reg_FL_ensemble.nc
- [ ] monarch_D3reg_FL_ensemble_localization.nc
- [x] monarch_D4reg_20220701_ensemble.nc
- [ ] monarch_D4reg_20220701_ensemble_localization.nc
- [x] monarch_D4reg_FL_20220701_ensemble.nc
- [ ] monarch_D4reg_FL_20220701_ensemble_localization.nc
- [x] monarch_3densvar_incr_2022070112.nc
- [x] monarch_3densvar_analysis_2022070112.nc
- [x] monarch_3densvar_analysis_lt_2022070112.nc
- [x] monarch_4densvar_incr_20220701.nc
- [x] monarch_4densvar_analysis_20220701.nc
- [x] monarch_4densvar_lt_analysis_20220701.nc  

Therefore the ensemble_localization files are missing their attributes.  
Strangely enough, the "single date" output of the 4D Dirac is missing the attributes, while the full 4D Dirac output is ok.
Of course not: the "single date" is harmless. Is the "ID" central block which breaks the attributes. We've probably missed a Fields initialization flavour.

andreapiacentini commented:
In my understanding, `dxi` is a pure increment, built from configuration and a vars list. It has no connection with a state and its metadata() whatsoever.
```
Increment4D_ dxi(geom, vars, xx.times(), *commTime);
```
When the Dirac output is obtained via an ensemble of States, `dxo` inherits the State metadata from the statistics on the StateSet.  
But when it comes from ID it is a simple copy of `dxi`, therefore with no access to metadata().  
For the same reason, the result of
```
      // Define output increment
      Increment4D_ dxo(dxi);

      // Setup localization
      Localization_ Lmat(geom, vars, locConfig);

      // Apply localization
      Lmat.multiply(dxo);
```
never gets in touch with any State metadata().   
Having the attributes in the Localization output is not useless: in particular for rotated grids they contain the recipe for display (e.g. ncview  or panoply use the `grid_mapping` and `coordinates` attributes for geographical representation). 

Pretty hard to figure out how to make the attributes become a "footprint" for every sort of output. Probably, my idea of storing them with every State isn't so good. I thought it safer, in case of different set of files (I don't know, some multiresolution thingy), but I should probably assume a stricter coherence in the files manipulated by BSC applications and keep this information as a permanent property of the IO factory (if somehow possible).  
Food for thought.

benjaminmenetrier commented:
I agree, it might be easier to generate and store the attributes the first time you read a file, and use them later. We could even do something smarter and store one set of attributes per grid (easy to do with a `std::unordered_map` and the grid UID).

andreapiacentini commented:
Midnight idea (or patch): 
in `saber/src/saber/oop/ErrorCovarianceToolbox.h`
I added the `dxi.diff` line in 
```
    if (diracParams != boost::none) {
      // Setup Dirac field
      Increment4D_ dxi(geom, vars, xx.times(), *commTime);
      dxi.diff(xx,xx);
      dxi.dirac(*diracParams);
      oops::Log::test() << "Input Dirac increment:" << dxi << std::endl;
```
and I've got all my attributes.

benjaminmenetrier commented:
> Midnight idea (or patch): in `saber/src/saber/oop/ErrorCovarianceToolbox.h` I added the `dxi.diff` line in
> 
> ```
>     if (diracParams != boost::none) {
>       // Setup Dirac field
>       Increment4D_ dxi(geom, vars, xx.times(), *commTime);
>       dxi.diff(xx,xx);
>       dxi.dirac(*diracParams);
>       oops::Log::test() << "Input Dirac increment:" << dxi << std::endl;
> ```
> 
> and I've got all my attributes.

I admire the midnight breakthrough, but I'm afraid it is not a robust solution for the future. Since the `Increment` constructor without metadata exists, it is not possible to ensure that nobody will use it one day. I would prefer the solution proposed in  #10.

andreapiacentini commented:
Superseeded by PR #10
